### PR TITLE
Implement end-to-end EKS integration test

### PR DESCRIPTION
**Pull request by fredjywang**
_Implementing feature requested in Issue [#100](https://github.com/aws/aws-cloud-map-mcs-controller-for-k8s/issues/100): End-to-end integration test using EKS._

*Description of changes:*
Using EKS, this new end-to-end integration test covers service exports and imports across two EKS clusters, as well as testing service consumption by the importing cluster. The test is largely based on [this](https://blog.bytequalia.com/kubernetes-multi-cluster-service-discovery-using-the-aws-cloud-map-mcs-controller/) blog. The test can be run using the command `make eks-integration-suite`.

This gives the project an integration test that more accurately represents the actual use-case of the controller, in that it tests the controller's functionalities with two EKS clusters instead of one local cluster, as the previous test had been. Adding this to the project workflows allow for more accurate testing of the current state of the controller, ensuring its core functionalities are maintained across commits.

*Testing:*

- The original integration test,`make integration-suite`, still runs successfully.
- The test has been successfully run multiple times in a row with no manual intervention
- Should the test fail for whatever reason, the user is able to run `make eks-integration-cleanup` and the EKS clusters are properly cleaned for any upcoming tests

*Next steps:*

- This integration test needs be added to project workflow
- The test requires the user to manually setup EKS clusters, a guide which will be made available to the public
- In the future, possible improvements to the test can be making it a one-touch test, so the user does not have to manually create the EKS clusters

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by astaticvoid — CHANGES_REQUESTED** [review]

File: config/controller_install_integration_test/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,7 @@
+bases:
+- ../default
+
+images:
+- name: controller
+  newName: ghcr.io/aws/aws-cloud-map-mcs-controller-for-k8s # update this to whatever image i want
```
[the remark this answers is not in the document]
Comment (fredjywang) [reply config/controller_install_integration_test/kustomization.yaml:6]: was meant to be for docker image deployment testing, will delete from PR as it is not used in integration test

File: integration/eks-test/scripts/eks-DNS-test.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env bash
+
+# Testing service consumption with client-hello pod
+
+source $1
+
+if [ "$#" -ne 1 ]; then
+    echo "test script expects common.sh as argument"
```
[the remark this answers is not in the document]
Comment (fredjywang) [reply integration/eks-test/scripts/eks-DNS-test.sh:8]: I saw this previously written in some other bash scripts, can remove if not necessary.

File: Makefile [hunk 3]
```diff
@@ -79,20 +79,36 @@ test-setup: ## setup test environment
 	test -f ${ENVTEST_ASSETS_DIR}/setup-envtest.sh || curl -sSLo ${ENVTEST_ASSETS_DIR}/setup-envtest.sh https://raw.githubusercontent.com/kubernetes-sigs/controller-runtime/v0.8.3/hack/setup-envtest.sh
 	source ${ENVTEST_ASSETS_DIR}/setup-envtest.sh; fetch_envtest_tools $(ENVTEST_ASSETS_DIR)
 
-integration-suite: ## Provision and run integration tests with cleanup
-	make integration-setup && \
-	make integration-run && \
-	make integration-cleanup
+kind-integration-suite: ## Provision and run integration tests with cleanup
```
Comment (astaticvoid) [inline Makefile:82]: workflow task names needs to be updated.
https://github.com/aws/aws-cloud-map-mcs-controller-for-k8s/blob/main/.github/workflows/integration-test.yml#L33

**Review by astaticvoid — CHANGES_REQUESTED** [review]

File: integration/eks-test/scripts/eks-cleanup.sh [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+source ./integration/eks-test/scripts/eks-common.sh
+
+# Delete service and namespace from cluster 1 & 2
+$KUBECTL_BIN config use-context $EXPORT_CLS
+$KUBECTL_BIN delete svc $SERVICE -n $NAMESPACE
+
+# Verfication to check if there are hanging ServiceExport or ServiceImport CRDs and clears the finalizers to allow cleanup process to continue
+for CRD in $($KUBECTL_BIN get crd -n $NAMESPACE | grep multicluster | cut -d " " -f 1 | xargs); do 
+    $KUBECTL_BIN patch crd -n $NAMESPACE $CRD --type merge -p '{"metadata":{"finalizers": [null]}}';
```
Comment (runakash) [inline integration/eks-test/scripts/eks-cleanup.sh:13]: Was this the problem with eks cleanup?

File: integration/eks-test/scripts/eks-common.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
+
+export KIND_BIN='./bin/kind'
+export KUBECTL_BIN='./testbin/bin/kubectl'
+export LOGS='./integration/eks-test/testlog'
```
Comment (runakash) [inline integration/eks-test/scripts/eks-common.sh:5]: LOGS should be dumped into /tmp/ folder in the server.

**Review by runakash — COMMENTED** [review]
Looks good overall. Two comments
* Adding more comments before each step explaining the step will go long way for the next developer.
* You may need to update the `integration/scenarios` path in some other places.

File: integration/eks-test/scripts/eks-cleanup.sh (hunk 4 above)
Comment (fredjywang) [reply integration/eks-test/scripts/eks-cleanup.sh:13]: Yes, leftover finalizers were making the CRDs unable to be properly cleaned up, and so the solution was to remove the finalizers, then remove the CRDs.

File: integration/eks-test/configs/client-hello.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: client-hello
+  namespace: aws-cloud-map-mcs-eks-e2e
+spec:
+  containers:
+    - command:
+        - sleep
+        - "1d"
```
[the remark this answers is not in the document]
Comment (fredjywang) [reply integration/eks-test/configs/client-hello.yaml:10]: This was taken from the original blog, so I kept it as is.

File: Makefile [hunk 7]
```diff
@@ -79,20 +79,36 @@ test-setup: ## setup test environment
...
+	make eks-integration-run && \
+	make eks-integration-cleanup
+
+eks-integration-setup: build kind test-setup ## Setup the integration test using EKS clusters
+	@./integration/eks-test/scripts/eks-setup.sh
+
+eks-integration-run: ## Run the integration test controller
+	@./integration/eks-test/scripts/eks-run-tests.sh
+
+eks-integration-cleanup: kind  ## Cleanup integration test resources in Cloud Map and EKS cluster
```
[the remark this answers is not in the document]
Comment (fredjywang) [reply Makefile:107]: Missed this; yes, `kind` is not needed here.

**Review by astaticvoid — DISMISSED** [review]

**Comment by astaticvoid** [comment]
No blockers for me, thanks for the improvements.

**Review by runakash — DISMISSED** [review]

**Comment by fredjywang** [comment]
The change to local `kubectl` usage was because the version in the testbin folder is 1.19.2, and caused some issues when trying to install and uninstall the controller. Local `kubectl` will be used for now to resolve this issue.

**Review by astaticvoid — APPROVED** [review]

File: integration/eks-test/scripts/eks-DNS-test.sh [hunk 8]
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env bash
+
+# Testing service consumption with client-hello pod
+
+echo "verifying cross-cluster service consumption..."
+
+kubectl exec $CLIENT_POD -n $NAMESPACE /bin/sh -- curl --version &>/dev/null
```
Comment (astaticvoid) [inline integration/eks-test/scripts/eks-DNS-test.sh:7]: I think it actually makes more sense to use $KUBECTL_BIN, and point it at local kubectl in eks-common.sh, so if we change to a distributed version we only need to change the path in one place.

**Review by astaticvoid — CHANGES_REQUESTED** [review]

File: integration/eks-test/scripts/eks-DNS-test.sh (hunk 8 above)
Comment (fredjywang) [reply integration/eks-test/scripts/eks-DNS-test.sh:7]: Makes sense, will add change.

**Pull request merged**